Add utility for downloading local data

mduan opened the pull request.

tdumitrescu commented on lib/report/util.js:
```diff
@@ -39,6 +39,33 @@ export function downloadData(endpoint, params, method=`POST`) {
+  // launch named download via hidden link
+  const link = document.createElement(`a`);
+
+  const downloadAttrSupported = (`download` in link);
```
superfluous parens

tdumitrescu commented on lib/report/util.js:
```diff
@@ -39,6 +39,33 @@ export function downloadData(endpoint, params, method=`POST`) {
+    // On Safari, it downloads the file with the name 'Unknown'
+    window.open(`data:attachment/csv;charset=utf-8,` + encodeURIComponent(dataStr));
+    return;
+  }
```
I'd prefer replacing this early return with a simple two-alternative conditional:
```js
if (downloadAttrSupported) {
  // ...
} else {
  // workaround for lesser browsers. drawbacks: ...
}
```

tdumitrescu commented on lib/report/util.js:
```diff
@@ -39,6 +39,33 @@ export function downloadData(endpoint, params, method=`POST`) {
+  // prepare blob
+  const blob = new Blob([dataStr], {type: `text/csv;charset=utf-8`});
+  const blobURL = URL.createObjectURL(blob);
+
```
since you left the comments preceding the other paragraphs, might as well retain this one's too:
```js
// launch named download via hidden link
```

tdumitrescu commented on lib/report/util.js:
```diff
@@ -39,6 +39,33 @@ export function downloadData(endpoint, params, method=`POST`) {
+  }
+
+  // prepare blob
+  const blob = new Blob([dataStr], {type: `text/csv;charset=utf-8`});
```
if this is going to be hardcoded to CSV, then the function needs a more specific name (`downloadLocalCSVData`?)

mduan replied:
I'll have it just hardcoded to CSV for now. We can generalize this later if we need to support data formats.

tdumitrescu commented on lib/report/util.js:
```diff
@@ -39,6 +39,33 @@ export function downloadData(endpoint, params, method=`POST`) {
+
+  link.style.display = `none`;
+  link.href = blobURL;
+  link.download = `${filename}.csv`;
```
another csv assumption

tdumitrescu commented on lib/report/util.js:
```diff
@@ -39,6 +39,33 @@ export function downloadData(endpoint, params, method=`POST`) {
+  if (!downloadAttrSupported) {
+    // Work-around for Safari. One drawback of this method is that it doesn't support setting a filename.
+    // On Safari, it downloads the file with the name 'Unknown'
+    window.open(`data:attachment/csv;charset=utf-8,` + encodeURIComponent(dataStr));
```
csv here too

mduan commented:
@tdumitrescu I've addressed your comments. Let me know when you've created a new build with the changes, and I can update Signal and Insights to use it.
